### PR TITLE
precompilepkgs: make the circular dep warning clearer and more informative

### PR DESCRIPTION
**Pull request by IanButterworth**
Was e.g.
```
┌ Warning: Circular dependency detected. Precompilation will be skipped for:
│   Base.PkgId(Base.UUID("eb0c05c4-6780-5852-a67e-5d31d2970b9a"), "ArrayInterfaceTrackerExt")
│   Base.PkgId(Base.UUID("f517fe37-dbe3-4b94-8317-1923a5111588"), "Polyester")
│   Base.PkgId(Base.UUID("0d7ed370-da01-4f52-bd93-41d350b8b718"), "StaticArrayInterface")
│   Base.PkgId(Base.UUID("6a4ca0a5-0e36-4168-a932-d9be78d558f1"), "AcceleratedKernels")
│   Base.PkgId(Base.UUID("244f68ed-b92b-5712-87ae-6c617c41e16a"), "NNlibAMDGPUExt")
│   Base.PkgId(Base.UUID("06b0261c-7a9b-5753-9bdf-fd6840237b4a"), "StaticArrayInterfaceStaticArraysExt")
│   Base.PkgId(Base.UUID("21141c5a-9bdb-4563-92ae-f87d6854732e"), "AMDGPU")
│   Base.PkgId(Base.UUID("9f7883ad-71c0-57eb-9f7f-b5c9e6d3789c"), "Tracker")
└ @ Base.Precompilation precompilation.jl:511
```
Now
Screenshot 2024-11-21 at 11 20 50 PM

Thanks to @topolarity figuring out proper cycles tracking.

### CODE REVIEW

**Comment by topolarity** [comment]
Is it possible to print the cycle as a chain?

It's not clear to me whether or not I can just read this list from top to bottom to get the cycle chain, so making that more obvious seems helpful

**Comment by IanButterworth** [comment]
I don't immediately know if these are ordered meaningfully. You might be better suited to answer that.. 

Also, what if there are multiple circularities.

**Comment by topolarity** [comment]
We should probably backport to 1.10 too, since we expect that users will be hitting extension cycles a _lot_ more often on that release.

In 1.11 these cycles are supposed to be impossible, except for environment tearing (where you end up with a package graph that doesn't actually `] resolve`) and any implementation bugs (some of which were not fixed until.. yesterday 😉)

Still very good to have (esp. on 1.10), just want to be clear about whether this is a "normal" error any more

File: base/precompilation.jl [hunk 1]
```diff
@@ -564,7 +591,26 @@ function _precompilepkgs(pkgs::Vector{String},
...
+                elseif i < length(cycle)
+                    line = "  │" * " " ^(i - 2)
+                else
+                    line = "  └" * "─" ^(max(0, i - 3)) * " "
+                end
+                line = color_string(line, :light_black) * full_name(exts, pkg) * "\n"
+                cycle_str *= line
+            end
+            cycles_names *= cycle_str
+        end
```
Comment (topolarity) [inline base/precompilation.jl:609]: Making a note here, so this isn't forgotten (mostly by me):

This also needs to display any packages in `circular_deps` that weren't in any cycle (i.e. they depend on a dependency cycle, so they were excluded from pre-compile)

Comment (IanButterworth) [reply base/precompilation.jl:612]: I just updated it to show these. Example up top

File: base/precompilation.jl [hunk 2]
```diff
@@ -564,7 +591,35 @@ function _precompilepkgs(pkgs::Vector{String},
...
+        plural1 = length(cycles) > 1 ? "these cycles" : "this cycle"
+        plural2 = length(cycles) > 1 ? "cycles" : "cycle"
+        msg = """Circular dependency detected.
+        Precompilation will be skipped for dependencies all in $plural1:
+        $cycles_names"""
+        if !isempty(outer_deps)
+            msg *= "Precompilation will also be skipped for the following, which have depdenencies in the $plural2:\n"
```
Comment (topolarity) [inline base/precompilation.jl]:
```suggestion
            msg *= "Precompilation will also be skipped for the following, which have dependencies in the $plural2:\n"
```

File: base/precompilation.jl [hunk 3]
```diff
@@ -564,7 +591,35 @@ function _precompilepkgs(pkgs::Vector{String},
...
+                end
+                line = color_string(line, :light_black) * full_name(exts, pkg) * "\n"
+                cycle_str *= line
+            end
+            cycles_names *= cycle_str
+        end
+        plural1 = length(cycles) > 1 ? "these cycles" : "this cycle"
+        plural2 = length(cycles) > 1 ? "cycles" : "cycle"
+        msg = """Circular dependency detected.
+        Precompilation will be skipped for dependencies all in $plural1:
```
Comment (topolarity) [inline base/precompilation.jl:617]:
```suggestion
        Precompilation will be skipped for dependencies in $plural1:
```

File: base/precompilation.jl [hunk 4]
```diff
@@ -564,7 +591,35 @@ function _precompilepkgs(pkgs::Vector{String},
         end
     end
     if !isempty(circular_deps)
-        @warn """Circular dependency detected. Precompilation will be skipped for:\n  $(join(string.(circular_deps), "\n  "))"""
+        outer_deps = copy(circular_deps)
```
Comment (topolarity) [inline base/precompilation.jl:594]: This warning message should probably be outlined to a separate function

File: base/precompilation.jl [hunk 5]
```diff
@@ -564,7 +591,36 @@ function _precompilepkgs(pkgs::Vector{String},
...
+            cycles_names *= cycle_str
+        end
+        plural1 = length(cycles) > 1 ? "these cycles" : "this cycle"
+        plural2 = length(outer_deps) > 1 ? "have" : "has"
+        plural3 = length(cycles) > 1 ? "cycles" : "cycle"
+        msg = """Circular dependency detected.
+        Precompilation will be skipped for dependencies all in $plural1:
+        $cycles_names"""
+        if !isempty(outer_deps)
+            msg *= "Precompilation will also be skipped for the following, which $plural2 depdenencies in the $plural3:\n"
```
Comment (topolarity) [inline base/precompilation.jl:620]:
```suggestion
            msg *= "Precompilation will also be skipped for the following, which depend on the above $plural3:\n"
```

I find this wording clearer, but the original is fine too - just make sure to fix the spelling of `dependencies`

**Review by topolarity — APPROVED** [review]
Looking great! I think it's much clearer to users what's going on here now.

Thanks for iterating on this @IanButterworth

**Comment by topolarity** [comment]
Oh the cycle order does look a bit "backwards" to me

I would have expected the dependency of a package to be above it, not below

**Comment by IanButterworth** [comment]
> I would have expected the dependency of a package to be above it, not below

I think of it the other way. Matching the sentence order for "Foo depends on Bar, which depends on Baz"
```
Foo
 Bar
  Baz
```

**Comment by topolarity** [comment]
Hmm.. What if we reverse the indentation from greatest to smallest?
```
  Baz
 Bar
Foo
```

Does that align our intuitions?

File: base/precompilation.jl [hunk 6]
```diff
@@ -566,33 +610,52 @@ function _precompilepkgs(pkgs::Vector{String},
...
         return false
     end
+    # set of packages that depend on a cycle (either because they are
+    # a part of a cycle themselves or because they transitively depend
+    # on a package in some cycle)
+    circular_deps = Base.PkgId[]
     for pkg in keys(direct_deps)
-        if scan_pkg!(pkg, direct_deps)
+        @assert isempty(stack)
+	if scan_pkg!(pkg, direct_deps)
```
Comment (IanButterworth) [inline base/precompilation.jl:658]:
```suggestion
        if scan_pkg!(pkg, direct_deps)
```

**Review by topolarity — APPROVED** [review]
I still find the staircase weird here (and my first instinct is to read it "backwards"), but this is a big improvement and all the functionality looks ship-shape to me 

Good to merge, I think!

**Comment by topolarity** [comment]
1.10 backport is done, but we should follow-up with 1.11 soon

**Pull request merged**